xla/hlo/analysis/while_loop_analysis.cc

```diff
@@ -21,83 +21,89 @@ limitations under the License.
-
-  // All operands of `instr` must be either constants or of the form
-  //   get-tuple-element(gte_operand, tuple_idx)
-  // for the same value tuple_idx. We also support the case where GTE feeds a
-  // copy that is then used.
-  optional<int64_t> tuple_idx;
-  for (const HloInstruction* operand : instr->operands()) {
-    if (Match(operand, m::Constant())) {
-      continue;
-    }
-    auto possibly_gte_operand = operand;
-
-    if (operand->opcode() == HloOpcode::kCopy) {
-      possibly_gte_operand = operand->operand(0);
-    }
-
-    if (possibly_gte_operand->opcode() != HloOpcode::kGetTupleElement) {
-      return nullopt;
-    }
-
-    if (!Match(possibly_gte_operand,
-               m::GetTupleElement(m::Op().Is(gte_operand)))) {
-      return nullopt;
-    }
-
-    int64_t operand_tuple_idx = possibly_gte_operand->tuple_index();
-    // This is the first GTE we are seeing. Set tuple_idx.
-    if (!tuple_idx.has_value()) {
-      tuple_idx = operand_tuple_idx;
-    } else {
-      if (operand_tuple_idx != tuple_idx) {
-        return nullopt;
-      }
-    }
-  }
-  return tuple_idx;
+// If `out` is a function of a single value in the tuple `in` and has no other
+// dependence, i.e. if `out=f(gte(in))`, then this function will return the
+// unique get-tuple-element index for the dependence.
+//
+// For example, in the following HLO, this function will return `1`:
+//   in = (s32[], s32[], s32[]) tuple(a,b,c)
+//   gte.1 = get-tuple-element(in), index=1
+//   out = fusion(gte.1), ...
+std::optional<int64_t> GetUniqueGTEDependenceIndex(const HloInstruction* out,
+                                                   const HloInstruction* in) {
+  if (out->parent() != in->parent() || !in->shape().IsTuple()) {
+    return std::nullopt;
+  }
+
+  // Extracts `out` as a function of `in`. We expect only one
+  // `get-tuple-element` operation on the input so that we can uniquely which
+  // element of `in` does `out` depend on. The checks for this are after the
+  // extraction.
+  std::unique_ptr<HloModule> extracted = ExtractModule(
+      /*instruction=*/out, /*height=*/-1, /*extract_selector=*/
+      [in](const HloInstruction* inst) -> bool { return inst != in; },
+      /*replace_type_selector=*/
+      [](const HloInstruction* inst) -> ReplaceType {
+        return ReplaceType::kReplaceParam;
+      });
+  HloComputation* entry = extracted->entry_computation();
+  if (entry->num_parameters() != 1 ||
+      entry->parameter_instruction(0)->shape() != in->shape()) {
+    return std::nullopt;
+  }
+  HloInstruction* param = entry->parameter_instruction(0);
+  if (param->user_count() == 0) return nullopt;
+  if (llvm::any_of(param->users(), [](const HloInstruction* inst) -> bool {
+        return inst->opcode() != HloOpcode::kGetTupleElement;
+      })) {
+    return std::nullopt;
+  }
+  int64_t candidate_index = param->users()[0]->tuple_index();
+  if (llvm::any_of(param->users(),
```

Comment:
why are you using llvm library? use absl::c_any_of instead

Reply:
Thanks! Changed it.